### PR TITLE
Add Google Analytics config option

### PR DESCRIPTION
**Pull request by coolaj86**
Adds the option `GOOGLE_ANALYTICS_ID` which puts the GA script tag into the `head.tmpl` so that a person can see metrics about their projects.

This would be of value to branded projects, such as Gitea itself (once moved to Gitea).

The ability to track analytics is something I personally see as a core benefit to having a personal git platform.

### CODE REVIEW

**Review by techknowlogick — CHANGES_REQUESTED** [review]
Thank you for PR. This has been discussed in #36. Due to the many possible analytics options (matomo, google analytics, fathom, etc...) as well as google not keeping a safeguard of data lately, I am against adding first class support for GA, especially with #3051 which allows users to use whichever analytics provider they prefer.

**Comment by coolaj86** [comment]
It sounds like this comes down to an issues of project values.

If the project is anti-Google, I can understand that. I don't agree with the stance, but I can accept it.

However, if one of the project values is **ease-of-adoption**, then I think a better approach would be to consider which value is more important.

I'm not a big Google fan. I look forward to the day when I can use a self-hosted analytics solution. If there is such a solution, perhaps we should bake that in too.

However, just like making it easy to integrate with GitHub is critical to the project's success and taking an anti-GitHub stance (which I definitely have) would be counter-productive, I think that saying "we don't add _this_ because someone might want _that_ instead - users can just figure it out by digging through the code on their own" is likewise counter-productive as if to say "we don't want to add GitHub login because someone might want to add Google login, and users can just dig through the code and add it themselves".

Using incrementalism and the 80/20 rule, I'm strongly in favor of adding Google Analytics. I'd also be in favor of adding other platforms. My goal is to see even greater adoption for Gitea by having a better user experience that supports of the lowest-common-denominator of user.

Integration ≠ Endorsement

**Comment by coolaj86** [comment]
That said, I'm so happy to see #3051, because that's been bugging me for a while as I've gone through version changes.

Thanks and I hope to see more of that. :)

**Comment by lafriks** [comment]
This can already be done using custom templates.

It would be nice to have section in docs that would describe on how to add this using custom template.

**Comment by techknowlogick** [comment]
lolya. that's exactly why I made that PR, because before themes I had a custom template that I had to change everytime I upgraded.

Two reasons why I'm against including this:
1. In terms of this JS because it technically isn't opensource licensed we aren't able to "freely" (free as in libre) distribute it.
2. Google updates their embed for analytics occasionally, even now at I've had to update from their universal analytics to google tag manager in other projects (ones that aren't open source).

However that being said, I am only one of many maintainers, and if others feel ok with including this then an owner is free to "dismiss" my review, and I'd be fine with that.

**Comment by coolaj86** [comment]
> In terms of this JS because it technically isn't opensource licensed we aren't able to "freely" (free as in libre) distribute it.

meh. It's pretty common in practice and I'm pretty sure that Google appreciates it. I'd be surprised to find anything that says it isn't licensed for free distribution.

> Google updates their embed for analytics occasionally, even now at I've had to update from their universal analytics to google tag manager in other projects (ones that aren't open source).

Once every 5 to 10 years. Even facebook login has that issue.

**Comment by techknowlogick** [comment]
>  It's pretty common in practice and I'm pretty sure that Google appreciates it.

Yes, I agree 100%, and while Google is likely cool with it too, sadly FSF uses Gitea and other "free software organizations" and Gitea is at a specific "free" level and that allows some organizations to use this. As even if they don't use the google code it is still "non-free" code that would be distributed by the binary (@MTecknology can tell you the woes of trying to package Gitea for debian even with completely "free" software)

(general note: when I say free I mean free as in libre, also I'll see if I can dig up in the tickets a source for my comment re: "free" level, I think we are rated B, and the only forge to be rated A is gnu savannah lol)

I also second @lafriks general comment above that the projects docs can be improved in regards to custom templates.

Edit: some "free software" sticklers, even see our Giteas of slack webhooks as non-free, however because we coded these ourselves (likely these just came directly from gogs before fork to give gredit to where credit is due) the code is technically free even though it interacts with a non-free service. People are free to have their own opinion on what is free software (and I know some people will be upset with me for even saying that because of course their own PoV is the only correct opinion), and that this debate is entirely exhausting and doesn't need to happen, but sadly because I am (one of many) stewards of Gitea I need to care at least a little amount.

Sorry for my novel, I really do appreciate the hardwork that you've done, and if this were any other project I wouldn't care at all and hit merge without a single thought, but because of Gitea's adoption within the "free software community" I reluctantly need to think about these kinds of things.

**Comment by coolaj86** [comment]
That’s like saying that

```
var answer = 42
console.log(“Hello World!”);
```

isn’t free.

There are only so many ways to read documentation and interpret how to accomplish 3 sequential steps. You couldn’t copyright that. It’s public domain.

The link to proprietary code isn’t even active unless the user opts in.

**Comment by coolaj86** [comment]
Anyway, since you have time to bikeshed I assume you have time to review one of my other PRs? ;)

**Comment by techknowlogick** [comment]
> That’s like saying that [code snippet] isn’t free.

Some of the free software community feel there is a bit more nuance to in than that (see links below).

> The link to proprietary code isn’t even active unless the user opts in.

Sadly though, Gitea would still be distributing it due to it being in the binary, and some think that still counts. I am not one of those people, and I think the saying goes "preaching to the choir" to what you are saying, in that I agree with you. 

A couple links for you to review, while I review your PRs ;)
* https://www.gnu.org/software/repo-criteria.html
* https://www.gnu.org/philosophy/javascript-trap.html

I personally don't agree with everything in those links, however I am bound as a part of this project to ensure that I on a best effort attempt to ensure the "free"-ness of this project.

**Comment by techknowlogick** [comment]
Closing per above discussion. @coolaj86 while this wasn't merged we definitely appreciate all the work that you put into this project. I'll work on opening a new PR with the docs on how to add analytics.

**Pull request closed without merging**
